# IMPORTS

In [50]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt 

from scipy.stats import kurtosis

import glob
import os


A module that was compiled using NumPy 1.x cannot be run in
NumPy 2.0.2 as it may crash. To support both 1.x and 2.x
versions of NumPy, modules must be compiled with NumPy 2.0.
Some module may need to rebuild instead e.g. with 'pybind11>=2.12'.

If you are a user of the module, the easiest solution will be to
downgrade to 'numpy<2' or try to upgrade the affected module.
We expect that some modules will need time to support NumPy 2.

Traceback (most recent call last):  File "c:\Users\raees\anaconda3\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "c:\Users\raees\anaconda3\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "c:\Users\raees\anaconda3\lib\site-packages\ipykernel_launcher.py", line 17, in <module>
    app.launch_new_instance()
  File "c:\Users\raees\anaconda3\lib\site-packages\traitlets\config\application.py", line 846, in launch_instance
    app.start()
  File "c:\Users\raees\anaconda3\lib\site-pack

AttributeError: _ARRAY_API not found

ImportError: numpy.core.multiarray failed to import

# PROBLEM STATEMENT

- Problem Overview

- Main Objective

## DATA LOADING

In [23]:
import pandas as pd
try:
    excel_url = "https://adbdatathon.s3.af-south-1.amazonaws.com/Datathon/Vehicle_Data.xlsx"
    
    print(f"Attempting to read Excel file from URL: {excel_url}\n")
    
    # No special options needed for a direct public HTTPS link
    df_excel = pd.read_excel(excel_url)
    
    print("✅ Successfully read Excel file from URL!")
    print("First 5 rows of the Excel data:")
    print(df_excel.head())

except Exception as e:
    print(f"❌ Error reading Excel file: {e}")
try:
    excel_url = "https://adbdatathon.s3.af-south-1.amazonaws.com/Datathon/part-00590-tid-6661262080813627668-5b71b33c-8263-4f63-9ea0-b9e2e435cab4-130881-1.c000.snappy.parquet"
    
    print(f"Attempting to read Excel file from URL: {excel_url}\n")
    
    # No special options needed for a direct public HTTPS link
    df_parquet = pd.read_parquet(excel_url)
    
    print("✅ Successfully read Excel file from URL!")
    print("First 5 rows of the Excel data:")
    print(df_parquet.head())

except Exception as e:
    print(f"❌ Error reading parquet file: {e}")


Attempting to read Excel file from URL: https://adbdatathon.s3.af-south-1.amazonaws.com/Datathon/Vehicle_Data.xlsx

✅ Successfully read Excel file from URL!
First 5 rows of the Excel data:
              IMEI        Make             Model           Colour    Year  \
0  860305051123794  VOLKSWAGEN     VW 27X - POLO      WHITE / WIT  2018.0   
1  860305051124123     HYUNDAI              ATOS  SILVER / SILWER  2022.0   
2  860305051147801        FORD            RANGER      GREY / GRYS  2022.0   
3  860305051168823      SUZUKI            ERTIGA      WHITE / WIT  2023.0   
4  860305051198176  VOLKSWAGEN  VW 240-POLO VIVO  SILVER / SILWER  2011.0   

    Age  Gender  
0  38.0    Male  
1  46.0    Male  
2  45.0    Male  
3  64.0    Male  
4  40.0  Female  
Attempting to read Excel file from URL: https://adbdatathon.s3.af-south-1.amazonaws.com/Datathon/part-00590-tid-6661262080813627668-5b71b33c-8263-4f63-9ea0-b9e2e435cab4-130881-1.c000.snappy.parquet

✅ Successfully read Excel file from URL!


In [28]:
owner = df_excel.copy()
owner.head()

IMEI        Make             Model           Colour    Year  \
0  860305051123794  VOLKSWAGEN     VW 27X - POLO      WHITE / WIT  2018.0   
1  860305051124123     HYUNDAI              ATOS  SILVER / SILWER  2022.0   
2  860305051147801        FORD            RANGER      GREY / GRYS  2022.0   
3  860305051168823      SUZUKI            ERTIGA      WHITE / WIT  2023.0   
4  860305051198176  VOLKSWAGEN  VW 240-POLO VIVO  SILVER / SILWER  2011.0   

    Age  Gender  
0  38.0    Male  
1  46.0    Male  
2  45.0    Male  
3  64.0    Male  
4  40.0  Female

The above data shows details of different models of cars owned by individuals. 

In [29]:
pd.set_option('display.max_columns',500)
tracking = df_parquet.copy()
tracking.head()

sourceimei  battery_voltage_value  hardware_attached_gps_not_present  \
0  869716067485269              12.484994                                  1   
1  869716067485269              12.845138                                  1   
2  869716067485269              12.845138                                  1   
3  869716067485269              12.845138                                  1   
4  869716067485269              12.845138                                  1   

             unitdatetime eventid   latitude  longitude  altitude  speed  \
0  2025-10-08 11:24:34 PM      5E -26.122560  28.078880    1639.0      0   
1  2025-10-08 03:59:49 PM      02 -26.122560  28.078880    1639.0      0   
2  2025-10-08 03:59:50 PM      02 -26.122560  28.078880    1639.0      0   
3  2025-10-08 03:59:51 PM      02 -26.122564  28.078883    1639.0      0   
4  2025-10-08 03:59:51 PM      00 -26.122560  28.078880    1639.0      0   

   heading  bearing  odometer  gforce_forward  gforce_backward  gforce_up  \
0       80       80    261911               0               50          0   
1       80       80    261911               0               50          0   
2       80       80    261911               0               50          0   
3       80       80    261911               0               50          0   
4       80       80    261911               0               50          0   

   gforce_down  gforce_left  gforce_right engine_hours  classification  \
0          950            0             0       159382              70   
1          950            0             0       159381              70   
2          950            0             0       159381              70   
3          950            0             0       159382              70   
4          950            0             0       159382              70   

  closest_suburb  condition       country  distance_suburb  \
0           None          1  South Africa                0   
1           None          1  South Africa                0   
2           None          1  South Africa                0   
3           None          1  South Africa                0   
4           None          1  South Africa                0   

           municipality postal_code province road  road_distance_m  \
0  City of Johannesburg        2090  Gauteng   ""                0   
1  City of Johannesburg        2090  Gauteng   ""                0   
2  City of Johannesburg        2090  Gauteng   ""                0   
3  City of Johannesburg        2090  Gauteng   ""                0   
4  City of Johannesburg        2090  Gauteng   ""                0   

   rough_road  roadspeed   suburb          town  
0           0          0  Bramley  Johannesburg  
1           0          0  Bramley  Johannesburg  
2           0          0  Bramley  Johannesburg  
3           0          0  Bramley  Johannesburg  
4           0          0  Bramley  Johannesburg

The above data shows vehicle tracking based on location, speed, vehicle components

In [26]:
df_excel.shape

(949, 7)

This dataset contains information on car makes and models and has 949 rows and 7 columns. 

In [27]:
df_parquet.shape

(58426, 33)

This dataset containing car telemetrics has 58 426 rows and 33 columns. 

## DATA PREPROCESSING AND CLEANING

### OWNER DATA

In [30]:
owner.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 949 entries, 0 to 948
Data columns (total 7 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   IMEI    949 non-null    int64  
 1   Make    948 non-null    object 
 2   Model   948 non-null    object 
 3   Colour  948 non-null    object 
 4   Year    948 non-null    float64
 5   Age     921 non-null    float64
 6   Gender  921 non-null    object 
dtypes: float64(2), int64(1), object(4)
memory usage: 52.0+ KB


The above shows a mixture of categorical and numerical data entries. Additionally, not all entries are non-null, which implies the presence of missing/null values.

In [31]:
owner.describe()

IMEI         Year         Age
count  9.490000e+02   948.000000  921.000000
mean   8.649012e+14  2018.279536   47.857763
std    4.276071e+12     4.627511   10.896567
min    8.603051e+14  2003.000000   26.000000
25%    8.603051e+14  2015.000000   39.000000
50%    8.647141e+14  2019.000000   47.000000
75%    8.697161e+14  2022.000000   56.000000
max    8.697161e+14  2024.000000   74.000000

- The IMEI is only an identifier and has no statistical relevance.
- From the data, car models range from oldest (2003) to newest (2024).
- The minimum age is 26 which is above the legal driving age in South Africa.

##### MISSING/NULL VALUES

In [33]:
# The implicance of missing/null values is there, How many though?
# How do we handle them?

owner.isnull().sum()

IMEI       0
Make       1
Model      1
Colour     1
Year       1
Age       28
Gender    28
dtype: int64

- The number of missing/null values across the features are less than 3%.
- We want to do a hypothesis testing for the randomness of the missing data (exploring the relation between age and gender)

In [ ]:
# Is it gender absent and age present, age absent and gender present or both absent?

null_owner_gender_age = owner[owner[['Age', 'Gender']].isnull().all(axis=1)]

null_owner_gender_age


IMEI               Make                                Model  \
59   864714060587087             TOYOTA                                HILUX   
74   868440060701058         VOLKSWAGEN                                 POLO   
183  864714060571628         VOLKSWAGEN                         VN 750-KOMBI   
186  864714060588960             SUZUKI                        VITARA BREZZA   
187  864714060615706                NaN                                  NaN   
215  869716064220305              MAZDA                          MAZDA CX-30   
228  869716066129082             NISSAN                                NP200   
261  860305056202742         VOLKSWAGEN                        VW 250 - POLO   
262  860305056227038                GWM                STEED 5 2.0 S S/C P/U   
282  860305056861265            RENAULT                                 KWID   
329  869716066084477                BMW                         X5 xDRIVE30d   
438  868440060817193             SUZUKI                             S-PRESSO   
439  869716064070023               HINO                200 310 (FH2) 4X2 F/C   
544  869716064199301         VOLKSWAGEN                     VW 241-POLO VIVO   
552  869716066116808  GREAT WALL MOTORS                             P SERIES   
556  869716066185225             NISSAN                       NP300 HARDBODY   
563  869716067469719         VOLKSWAGEN                             POLO 1.6   
590  860305055025409                BMW                                   M2   
634  864714060618445            HYUNDAI                            GRAND i10   
652  868440060710190             NISSAN                                NV350   
654  868440060766275            HYUNDAI                         GRAND I10 BA   
699  860305051214031             TOYOTA  HILUX 2.8 GD-6 GR-S 4X4 A/T P/U D/C   
719  860305056271390               FORD                               RANGER   
720  860305056273099             NISSAN                                NP200   
744  860305056781646             TOYOTA                                HILUX   
772  868440060706065             SUZUKI                                DZIRE   
903  868440060848461         MITSUBISHI                  ASX 2.0 5DR GLS A/T   
940  869716068376038              ISUZU     DMAX 250 HO HI-RIDER A/T D/C P/U   

              Colour    Year  Age Gender  
59             WHITE  2023.0  NaN    NaN  
74             WHITE  2021.0  NaN    NaN  
183      WHITE / WIT  2011.0  NaN    NaN  
186  ORANGE / ORANJE  2023.0  NaN    NaN  
187              NaN     NaN  NaN    NaN  
215       RED / ROOI  2023.0  NaN    NaN  
228            WHITE  2023.0  NaN    NaN  
261      GREY / GRYS  2023.0  NaN    NaN  
262            WHITE  2023.0  NaN    NaN  
282            WHITE  2023.0  NaN    NaN  
329            WHITE  2023.0  NaN    NaN  
438           SILVER  2023.0  NaN    NaN  
439            WHITE  2023.0  NaN    NaN  
544            WHITE  2023.0  NaN    NaN  
552            WHITE  2023.0  NaN    NaN  
556            WHITE  2023.0  NaN    NaN  
563           SILVER  2019.0  NaN    NaN  
590            BLACK  2018.0  NaN    NaN  
634            WHITE  2023.0  NaN    NaN  
652            WHITE  2023.0  NaN    NaN  
654             BLUE  2018.0  NaN    NaN  
699            BLACK  2023.0  NaN    NaN  
719            WHITE  2023.0  NaN    NaN  
720            WHITE  2023.0  NaN    NaN  
744            WHITE  2023.0  NaN    NaN  
772            WHITE  2023.0  NaN    NaN  
903           SILVER  2023.0  NaN    NaN  
940            WHITE  2024.0  NaN    NaN

- All the missing/null  values for both features are simultaneously null. This may imply that the missing/null values are not missing at random as individuals might not have been comfortable with disclosing their personal information such as age and gender. 
- As the unique entries of age and gender cannot be assumed, we have decided to drop the missing/null values.

In [42]:
owner = owner.dropna()

owner.isnull().sum()


IMEI      0
Make      0
Model     0
Colour    0
Year      0
Age       0
Gender    0
dtype: int64

Removed/Dropped missing value entries.

In [44]:
# Duplicates
print(owner.duplicated().sum())

0


There are no duplicates

#### OUTLIERS

### TRACKING DATA